lib/config.js

```diff
@@ -1,14 +1,43 @@
+'use strict';
+
 var config = {
-    brand: 'Discord',
-    token: process.env.OAUTH_TOKEN,
-    trackingID: process.env.TRACKING_ID,
-    commentWait: process.env.COMMENT_WAIT || 250,
-    commentAttempts: process.env.COMMENT_ATTEMPTS || 10,
-
-    // Used when no environment variable (process.env.PORT) is set
-    port: 3000,
-    host: 'localhost',
-    protocol: 'http:'
+    def: {
+        brand: 'Discord',
+        token: process.env.OAUTH_TOKEN,
+        commentAttempts: process.env.COMMENT_ATTEMPTS || 10,
+    },
+    development: {
+        port: 3000,
+        redisURL: 'redis://localhost',
+        databaseURL: 'postgres://localhost/discord_test'
+    },
+    test: {
+        protocol: 'http:',
+        host: 'localhost',
+        port: 3000,
+        trackingID: '654321',
+        commentWait: 0,
+        redisURL: 'redis://localhost',
+        databaseURL: 'postgres://localhost/discord_test',
+        database: 'discord_test'
+    },
+    production: {
+        port: process.env.PORT,
+        trackingID: process.env.TRACKING_ID,
+        commentWait: process.env.COMMENT_WAIT || 250,
+        redisURL: process.env.REDIS_URL,
+        databaseURL: process.env.DATABASE_URL
+    }
```

Comment:
Really good points.

Putting default values straight into `lib/config.js` is, IMO, fine. In fact, I use `python-decouple` for python 12factor config, which enforces a `default=` argument that's then overridden by an environment variable, if it finds one. [See codesy's `settings.py` for example](https://github.com/codesy/codesy/blob/master/codesy/settings.py#L23-L27). So, its precedence is environment variables first, then falling back to a default.

`.env` is just a convenience file for keeping the development environment values in the repo without violating the 12factor principle. I.e., the application doesn't actually use `.env` - it uses the environment variables that `.env` and `autoenv` publishes for the app. It's also common to include a `.env-dist` file in the repo, and copy it over to `.env` as part of setting up the local app.